webapp/src/dogma/features/auth/authSlice.ts

```diff
@@ -16,130 +16,141 @@
+    }
+  }
 });
 
 export interface LoginParams {
   username: string;
   password: string;
 }
 
-export const login = createAsyncThunk('/auth/login', async (params: LoginParams, thunkAPI) => {
-  const response = await axios.post(`${process.env.NEXT_PUBLIC_HOST}/api/v1/login`, qs.stringify(params), {
-    validateStatus: (status) => status < 500,
-  });
-  if (response.status === HttpStatusCode.Ok) {
-    setSessionId(response.data.access_token);
-    await thunkAPI.dispatch(getUser());
-    return true;
-  }
-
-  // TODO(ikhoon): Replace alert with Modal
-  alert('Cannot sign in Central Dogma web console. Please check your account and password again.');
-  return false;
-});
-
-export const logout = createAsyncThunk('/auth/logout', async () => {
-  const response = await axios.post(`${process.env.NEXT_PUBLIC_HOST}/api/v1/logout`, null, {
-    validateStatus: (status) => status < 500,
-  });
-  if (response.status === HttpStatusCode.Ok) {
-    removeSessionId();
-    return true;
+export const login = createAsyncThunk('/auth/login', async (params: LoginParams, { rejectWithValue }) => {
+  try {
+    const config = {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+    };
+    const { data } = await axios.post(`${process.env.NEXT_PUBLIC_HOST}/api/v1/login`, params, config);
+    return data;
+  } catch (error) {
+    // TODO(ikhoon): Replace alert with Modal
+    alert('Cannot sign in Central Dogma web console. Please check your account and password again.');
+    if (error.response && error.response.data.message) {
+      return rejectWithValue(error.response.data.message);
+    } else {
+      return rejectWithValue(error.message);
+    }
   }
-  alert('Problem logging out. Please try again.');
-  return false;
 });
 
 interface UserSessionResponse {
   isAuthorized: boolean;
   user?: UserDto;
 }
 
-export const validateSession = createAsyncThunk<UserSessionResponse>('/auth/validate_session', async () => {
-  const response = await axios.get(`${process.env.NEXT_PUBLIC_HOST}/security_enabled`, {
-    validateStatus: (status) => status < 500,
-  });
-  if (response.status === HttpStatusCode.NotFound) {
-    // Anonymous mode
-    removeSessionId();
-    return { isAuthorized: true };
-  }
-
-  if (response.status === HttpStatusCode.Ok) {
-    if (getSessionId() === null) {
-      return { isAuthorized: false };
-    }
-
-    const userResponse = await axios.get(`${process.env.NEXT_PUBLIC_HOST}/api/v0/users/me`, {
-      validateStatus: (status) => status < 500,
-    });
-
-    if (userResponse.status === HttpStatusCode.Ok) {
-      // The current session ID is valid. Login is not required.
-      return { isAuthorized: true, user: userResponse.data as UserDto };
+export const validateSession = createAsyncThunk<UserSessionResponse>(
+  '/auth/validate_session',
+  async (_, { rejectWithValue }) => {
+    try {
+      await axios.get(`${process.env.NEXT_PUBLIC_HOST}/security_enabled`);
```

Comment:
I don't see `security_enabled` can validate the current `sessionId`. It does not check the `sessionId` and returns 200 or 404 status.
`/api/v0/users/me` was added to check the current session is valid. If the API returns `401 Unauthorized`, we can assume the the session expires.

Reply:
`security_enabled` is only checking the mode (anonymous or not). 

The validation happens in the next step. 
```
.addCase(getUser.rejected, (state) => {
        state.user = null;
        state.sessionId = '';
        if (typeof window !== 'undefined') {
          localStorage.removeItem('sessionId');
        }
      });
```
Here if `getUser` fails then `sessionId` gets cleared.

Reply:
Seems like we use this pattern to sequentially execute two thunks. https://stackoverflow.com/a/67028881/1736581

Reply:
> It looks good but I'm worried accessing localStorage might validate the basic rule of reducers.

Persisting the store is actually a common case, there's even a library for it https://www.npmjs.com/package/redux-persist

Reply:
> It looks good but I'm worried accessing `localStorage` might validate the basic rule of reducers.

Let me move them.

Reply:
>  Should we also fix the following code to access sessionId from the redux store? The second parameter of prepareHeaders allows us to call getState().

This comment is not addressed. Generally, it would work well. But I prefer to use one of `localStorage.getItem('sessionId')` and `state.auth.sessionId` for consistency. 
As we decided to add `sessionId` to `auth` store, it would be worth using `state.auth.sessionId` for reading.

- `localStorage.getItem('sessionId')` is only used to read the initial value.
- `localStorage.setItem()` and `localStorage.removeItem()` are handled in `createThunkAsync`.
- Rest of them, should we use `state.auth.sessionId`?

Reply:
> 



> I'm OK if we only use `localStorage` as an auxiliary storage to get initial values. Otherwise, I think `auth.sessionId` should be used outside of reducers. You have fixed all other code but one thing seems to remain. Should we also fix the following code to access `sessionId` from the redux store? The second parameter of `prepareHeaders` allows us to call `getState()`.
> 
> https://github.com/line/centraldogma/blob/5b1514a3cc2f61e6bbdec8fe7911ddd1297d681c/webapp/src/dogma/features/api/apiSlice.ts#L24-L25

Somehow I missed it as it was in a separate file. Thanks for pointing it out. Updated.